diff --git a/browser/components/urlbar/UrlbarMuxerStandard.sys.mjs b/browser/components/urlbar/UrlbarMuxerStandard.sys.mjs
index 1111111111111111111111111111111111111111..2222222222222222222222222222222222222222 100644
--- a/browser/components/urlbar/UrlbarMuxerStandard.sys.mjs
+++ b/browser/components/urlbar/UrlbarMuxerStandard.sys.mjs
@@ -920,6 +920,10 @@
       result.providerName == "UrlbarProviderHeuristicFallback" &&
       state.context.heuristicResult?.providerName !=
         "UrlbarProviderHeuristicFallback"
+      && !(
+        result.type == UrlbarUtils.RESULT_TYPE.SEARCH &&
+        state.context.heuristicResult?.type == UrlbarUtils.RESULT_TYPE.URL
+      )
     ) {
       return false;
     }
@@ -1582,7 +1586,9 @@
       usedLimits.maxResultCount++;
     }
 
+    if (!(result.heuristic && result.source == UrlbarUtils.RESULT_SOURCE.ZEN_ACTIONS)) {
     state.usedResultSpan += span;
+    }
     this._updateStatePostAdd(result, state);
 
     return true;
